XWIKI-16140: Make quick search suggestions accessible

Sereza7 opened the pull request:
**Jira:** https://jira.xwiki.org/browse/XWIKI-16140
**PR Changes:**
- Set the search form as the quick search modal parent. 
- Changed the clickable divs in the quick search menu to buttons. 
- Increased contrast on a .css muted-text (hide quick search)

**View:**
The DOM tree before (right) and after (left) this PR
![16140-UpdatedDOM](https://user-images.githubusercontent.com/28761965/230076021-a4db1af8-af27-40db-a361-8607043e7943.png)
Showcase of keyboard navigation:
https://up1.xwikisas.com/#1X2CBROxBl50VKazoiTyeg

manuelleduc commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/search/searchSuggest.css:
```diff
@@ -21,12 +21,12 @@ ul.xlist li.showAllResults.loading {
   border: 1px solid #EEEEEE;
   font-size: 1em;
-  height: 25px;
+  height:100%;
```
```suggestion
  height: 100%;
```

manuelleduc commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/search/searchSuggest.js:
```diff
@@ -131,19 +131,17 @@ var XWiki = (function (XWiki) {
-          'value' : valueNode
+          containerClasses: 'suggestItem',
+          classes: 'showAllResults',
+          eventCallbackScope : this,
+          noHighlight : true,
+          value : valueNode,
```
```suggestion
          eventCallbackScope: this,
          noHighlight: true,
          value: valueNode,
```

manuelleduc commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/suggest/suggest.css:
```diff
@@ -14,24 +14,32 @@ body {
-  margin: 0 4px;
-  opacity: .4;
-  padding: 0 0.5em;
+  opacity: .8; /* Should eventually use a full color defined in the color theme like .text-muted */
```
Isn't is possible to add the `text-muted` class to the `div.suggestItems .hide-button` elements?

Sereza7 replied:
It's not possible here since it's a CSS file and we don't have access to inheritance like we could do for LESS.

However I didn't think about it before, but we could just add the class `.text-muted` to the span `.hide-button`.
Is it correct to use a class just to apply style like this?
@manuelleduc

Sereza7 replied:
I think I can add a couple line to set the color of this element in the flamingo skin file: `action-menus.less`.

Sereza7 replied:
Done :+1:

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/importer/import.css:
```diff
@@ -156,15 +156,15 @@
-#import ul.package li.xitem div.xitemcontainer span.checkbox,
-#import ul.package li.xitem div.xitemcontainer input.space {
+#import ul.package li.xitem button.xitemcontainer span.checkbox,
+#import ul.package li.xitem button.xitemcontainer input.space {
```
I'm not convinced by the usage of the button element. We use the anchor element in other suggest drop downs (all suggest pickers) or menus (e.g. More Actions menu). It makes a lot of sense to be able to open a search suggest result in a new tab. You can't do this with the button.

Sereza7 commented:
**PR Changes:**
* Updated the style to take into account theme colors and avoid using opacity
* Replaced some buttons with anchors (the view all results submits a form, so I left it as a button).
* Proper management of highlight on focus

**New demo:** https://up1.xwikisas.com/#Am9TmRi-eF7ZeOHFUda0rw

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/less/action-menus.less:
```diff
@@ -149,6 +149,10 @@
     }
   }
+
+  div.suggestItems .hide-button{
```
```suggestion
  div.suggestItems .hide-button {
```

Also, I'm not fully sure about the location. I mean, I understand the need for LESS, but I'd put this rule maybe in ``misc.less`` since it's not related to action menus (the same search suggest can be used in many other places).

mflorea replied:
```suggestion
  div.suggestItems .hide-button {
```

Also, I'm not fully sure about the location. I mean, I understand the need for LESS, but I'd put this rule maybe in ``misc.less`` since it's not related to action menus (the same search suggest can be used in many other places).

mflorea commented on xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-ui/src/main/resources/XWiki/QuickSearchUIX.xml:
```diff
@@ -143,8 +143,9 @@
-        if (document.activeElement !== globalSearchButton[0] &amp;&amp; document.activeElement !== globalSearchInput[0]) {
+        if (!document.getElementById('globalsearch').contains(document.activeElement)) {
           globalSearch.addClass('globalsearch-close');
+          document.fire("xwiki:suggest:collapsed",globalSearchInput);
```
We're not supposed to use Prototype.js API any more, see http://api.prototypejs.org/dom/document/fire/index.html , especially inside a require block that gets jQuery injected. See https://api.jquery.com/trigger/#trigger-eventType-extraParameters .

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/search/searchSuggest.js:
```diff
@@ -85,10 +86,11 @@ var XWiki = (function (XWiki) {
         // Show the "No results!" message.
         this.noResultsMessage.removeClassName('hidden').setStyle({'float': 'left'});
       }
+      event.memo.container.select('.suggestItem')[0].href=event.memo.url||"";
```
```suggestion
      event.memo.container.select('.suggestItem')[0].href = event.memo.url || "";
```

Also, this deserves a comment because it's not obvious why you're changing the URL of the first suggest item that is supposed to point to the search page.

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/search/searchSuggest.js:
```diff
@@ -102,7 +104,7 @@ var XWiki = (function (XWiki) {
           Event.stop(event.memo.originalEvent);
         }
-        if (!event.memo.url) {
+        if (!event.memo.url || event.memo.url==="") {
```
The added condition is redundant. ``!'' === true``

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/search/searchSuggest.js:
```diff
@@ -131,29 +137,26 @@ var XWiki = (function (XWiki) {
         }
       });
       var allResults = allResultsNode.getElement();
+      allResultsNode.items[0].getElement().addEventListener('focusin', (event) => this.suggest.setHighlight($(event.currentTarget)));
```
Line too long (>120 chars).

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/suggest/suggest.js:
```diff
@@ -156,6 +156,8 @@ var XWiki = (function(XWiki){
       this.seps = "";
     }
 
+    this.searchPageUrl = encodeURIComponent(this.fld.value.strip());
```
```suggestion
    this.searchPageURL = encodeURIComponent(this.fld.value.strip());
```

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/suggest/suggest.js:
```diff
@@ -727,9 +729,11 @@ var XWiki = (function(XWiki){
+        containerNature: 'a'
       });
-
+      item.containerElement.setAttribute('href',arr[i].url);
```
```suggestion
      item.containerElement.setAttribute('href', arr[i].url);
```

mflorea commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/uicomponents/widgets/list/xlist.js:
```diff
@@ -98,7 +98,8 @@ var XWiki = function(XWiki){
           var classes = 'xitem ' + (this.options.noHighlight ? '' : 'xhighlight ');
           classes += this.options.classes ? this.options.classes: '';
-          this.containerElement = new Element("div", {'class': 'xitemcontainer'}).insert(content || '');
+          var containerNature = this.options.containerNature ? this.options.containerNature: 'div';
```
```suggestion
          var containerNature = this.options.containerNature || 'div';
```

But I'm not sure about the "Nature" part. I haven't seen this term used when talking about DOM and HTML. Maybe ``containerTagName`` is more explicit?

Sereza7 replied:
I changed it to `containerTagName` in the last commit :+1: 
I was not sure what to call it, I also thought about `containerType` but I couldn't really find a name for this string when looking online.
